framework/src/media/media_recorder.c

```diff
@@ -0,0 +1,342 @@
+	sem_t               sem_pause;
+	struct pcm         *pcmin;
+	struct pcm_config  *pcmconf;
+	unsigned int        buffer_size;
+	int                 fd;
+	media_format_e      format;
+	record_state_t      state;
+};
+
+static struct media_record_context_s g_rc;
+static bool b_started = false;
+
+int threadfunc_media_recording(void *args);
+
+/**
+ * Public Functions
+ */
+record_result_t media_record(int fd)
+{
+	pthread_attr_t attr;
+	struct sched_param sparam;
+
+	g_rc.fd = fd;
+
+	if (IS_SOCKET(g_rc.fd)) {
+		char instruction = 0x04;
+		send(g_rc.fd, &instruction, 1, 0);	
+	}
+	
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_STOPPING && g_rc.state != RECORD_AVAILABLE) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}
+
+	if (g_rc.state == RECORD_STOPPING) {
+		pthread_join(g_rc.pth, NULL);
+		g_rc.state = RECORD_AVAILABLE;
+	}
+
+	if (g_rc.state == RECORD_AVAILABLE) {
+		g_rc.fd = fd;
+		g_rc.thread_running = true;
+
+		if (pthread_attr_init(&attr) != 0) {
+			goto error_thread_init;
+		}
+
+		sparam.sched_priority = CONFIG_MEDIA_PRIORITY;
+		if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setstacksize(&attr, CONFIG_MEDIA_STACK_SIZE) != 0) {
+			goto error_thread_init;
+		}
+		if (pthread_create(&g_rc.pth, &attr, (pthread_startroutine_t)threadfunc_media_recording, NULL) != 0) {
+			goto error_thread_init;
+		}
+
+		g_rc.state = RECORD_RUNNING;
+	}
+
+	RECORD_UNLOCK();
+	return RECORD_OK;
+
+error_thread_init:
+	g_rc.thread_running = false;
+	RECORD_UNLOCK();
+	return RECORD_ERROR_THREAD_INIT;
+}
+
+record_result_t media_record_set_config(int channel, int sample_rate, int pcm_format, media_format_e format)
+{
+	RECORD_LOCK();
+
+	if (g_rc.pcmconf) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}
+
+	g_rc.pcmconf = (struct pcm_config *)malloc(sizeof(struct pcm_config));
+	if (!g_rc.pcmconf) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_MALLOC;
+	}
+
+	g_rc.pcmconf->channels = channel;
+	g_rc.pcmconf->rate = sample_rate;
+	g_rc.pcmconf->format = pcm_format;
+	g_rc.format = format;
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_record_prepare(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.pcmin) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}	
+
+	g_rc.pcmin = pcm_open(0, 0, PCM_IN, g_rc.pcmconf);
+	if (!g_rc.pcmin) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_DRIVER;
+	}
+
+	pcm_prepare(g_rc.pcmin);
+	g_rc.buffer_size = pcm_frames_to_bytes(g_rc.pcmin, pcm_get_buffer_size(g_rc.pcmin));
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+bool media_is_recording(void)
+{
+	return (g_rc.state == RECORD_RUNNING);
+}
+
+record_result_t media_stop_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state == RECORD_PAUSED) {
+		g_rc.state = RECORD_STOPPING;
+		sem_post(&g_rc.sem_pause);
+	} else if (g_rc.state != RECORD_RUNNING) {
```

Comment:
If pause and stop take some times,
we don't have to add STOPPED / PAUSING???

Reply:
Actually, I kept these names. Pause is a quick operation, it is just setting the state variable to paused. Stop operation takes some time so I named it stopping. I thought about STOPPED state but STOPPED is same as AVAILABLE state. So I omitted having the STOPPED state.

Reply:
alright =)